prov/sockets: new code for the connection manager for improved scalability

sydidelot opened the pull request:
Hi Libfabric's team!

I have been using Libfabric in my company (DDN Storage) for the last year.
I have already contributed from time to time to the Verbs/socket/PSM2 providers. Today I come up with a pretty long patch series that mainly focuses on improvements/fixes to the socket provider.

The general goal of this patch series is to fix scalability issues in the socket's connection manager.
With the current code, libfabric requires the creation of 1 POSIX thread and 2 file descriptors (socketpair, to signal any connection thread blocked in epoll_wait()) per connection.

This patch series completely rewrites this connection manager.
The general idea of the new code is the following: one connection manager (CM) thread is created per OFI domain, which is responsible for progressing connection establishments for the given domain.
The new design mimics the code from **socket_progress.c** where a network packet that has been partially progressed will be later resumed, so that it doesn't block the connection manager thread.

I know that most of you guys are currently busy with the upcoming 1.5.0 release and this is probably not the best moment to accept such huge patches.
However, I would appreciate some reviews/feedback on this code :-)

I'm open to discussions, suggestions or anything I can do to help you for the review.

I've been doing my best to test/fix the patches but there are still some issues:
- The Linux tests in Travis are failing from time to time. I'm still investigating this issue.
- The OSX tests in Travis are failing all the time (looks like **fi_cm_data hangs**). I don't have access to an OSX platform yet, I'll see what I can do.

Thanks again for your help,
Sylvain

shefty commented:
Thanks for the patch.  I will look it over when I get a chance.  Any improvement to the socket provider is welcome.

Along this line, I'm actually hoping that we can deprecate the socket provider in favor of using the ofi-rxm + tcp providers.  The tcp provider will still be a socket based provider, but only support the MSG endpoint.  The rxm provider will implement reliable-datagram semantics over msg endpoints.  The rxm code will be more generic in that it will also work over the verbs provider.  This will increase code sharing and simplify maintenance, so that improvements to rxm can work with all underlying providers.

Of course, this all depends on the combined rxm +tcp solution providing the same performance and scalability as sockets by itself.  Basically, if the socket code can leverage shared code, in particular that used by rxm, it could be better long term.  Or even feeding the same improvements into rxm would be great.  Or even just generalizing the socket code can be helpful.  (A lot of the existing utility code is generalized socket code.)

But don't interpret this to mean that we don't want to improve sockets!  Sockets is all we have today, and even our apps depend on it working well across medium sized clusters.

shefty commented on prov/sockets/src/sock_rx_entry.c:
```diff
@@ -72,7 +72,6 @@ struct sock_rx_entry *sock_rx_new_entry(struct sock_rx_ctx *rx_ctx)
 		entry = slist_remove_head(&rx_ctx->pool_list);
 		rx_entry = container_of(entry, struct sock_rx_entry, pool_entry);
 		rx_entry->rx_ctx = rx_ctx;
-		entry = slist_remove_head(&rx_ctx->pool_list);
```
Thanks - cherry picking this patch to merge manually.

sydidelot replied:
:+1:

shefty commented on include/fi_list.h:
```diff
@@ -103,6 +103,10 @@ static inline void dlist_remove(struct dlist_entry *item)
+#define dlist_foreach_safe(head, item, tmp) \
+    for ((item) = (head)->next, (tmp) = (item)->next; (item) != (head); \
+              (item) = (tmp), (tmp) = (item)->next)
+
```
thanks - will cherry pick to merge manually

sydidelot replied:
:+1:

shefty commented on src/fi_tostr.c:
```diff
@@ -690,44 +690,44 @@ char *DEFAULT_SYMVER_PRE(fi_tostr)(const void *data, enum fi_type datatype)
 		break;
 	case FI_TYPE_OP_TYPE:
-		fi_tostr_op_type(buf, enumval);
+		fi_tostr_op_type(buf, *enumval);
```
Thanks - I will cherry pick this and merge manually

sydidelot replied:
:+1:

shefty commented on prov/sockets/src/sock_progress.c:
```diff
@@ -152,6 +152,9 @@ static void sock_pe_release_entry(struct sock_pe *pe,
 	dlist_remove(&pe_entry->ctx_entry);
 
+	if (pe_entry->type == SOCK_PE_RX)
+		assert(ofi_rbavail(&pe_entry->comm_buf) == pe_entry->comm_buf.size);
```
Please rewrite as:
```
assert((pe_entry->type !=SOCK_PE_RX) || ofi_rbempty(&pe_entry->comm_buff));
```

shefty commented on prov/sockets/src/sock_progress.c:
```diff
@@ -169,6 +172,13 @@ static void sock_pe_release_entry(struct sock_pe *pe,
+		pe_entry->comm_buf.rcnt =
+			pe_entry->comm_buf.wcnt =
+			pe_entry->comm_buf.wpos = 0;
+	}
```
Please add a new ring buffer call for this, ofi_rbreset, or call ofi_rbdiscard(rb, ofi_rbused(rb))

shefty commented on prov/sockets/src/sock_progress.c:
```diff
@@ -220,6 +230,8 @@ static struct sock_pe_entry *sock_pe_acquire_entry(struct sock_pe *pe)
 		entry = pe->free_list.next;
 		pe_entry = container_of(entry, struct sock_pe_entry, entry);
+		/* Ring buffer must be empty */
+		assert(ofi_rbavail(&pe_entry->comm_buf) == pe_entry->comm_buf.size);
```
Use ofi_rbempty()

shefty replied:
I ended up pulling in this patch manually, and made the above mentioned changes.

shefty commented on prov/sockets/src/sock_ep.c:
```diff
@@ -1732,6 +1737,8 @@ int sock_alloc_endpoint(struct fid_domain *domain, struct fi_info *info,
 		tx_ctx->ep_attr = sock_ep->attr;
 		tx_ctx->domain = sock_dom;
+		if (tx_ctx->rx_ctrl_ctx && tx_ctx->rx_ctrl_ctx->is_ctrl_ctx)
+			tx_ctx->rx_ctrl_ctx->domain = sock_dom;
```
Blah - the current code is a mess in this area.  For now, I think it would help if the changes from this patch were squished into sock_tx_alloc_ctx().  That appears to be the only place where the domain isn't initialized. (You will need to pass the domain through to that call.)

All of this should really have proper reference counting, but that's a bigger change than I want to deal with for now.

sydidelot replied:
I'll modify sock_tx_alloc_ctx() to pass the domain in arguments. Thanks!

shefty replied:
FYI- I pulled this patch in as-is, since it provides a real fix.

shefty commented:
See #3195 - bug fix patches selected to be upstreamed for the 1.5 release.

shefty commented on prov/sockets/src/sock_progress.c:
```diff
@@ -1907,6 +1907,8 @@ static int sock_pe_progress_tx_entry(struct sock_pe *pe,
 
 		sock_pe_report_tx_error(pe_entry, 0, FI_EIO);
+		pe_entry->is_complete = 1;
+
```
thanks - cherry picked this patch

sydidelot replied:
:+1:

shefty commented on prov/sockets/include/sock.h:
```diff
@@ -869,6 +869,8 @@ struct sock_pe_entry {
 	struct ofi_ringbuf comm_buf;
 	size_t cache_sz;
+
+	ofi_atomic32_t completion_reported;
```
I'm hesitant to add yet another state variable to the pe_entry.  We already have is_complete, is_error, and mr_checked.  There should be some way we can combine these into a single state machine.  is_complete is particularly odd, since I can't easily figure out what it means.  :/  It's set to 1 in several places after a completion is reported, but not all places.  We also need to ensure that completions aren't being reported multiple times when counters are used, and not just in the cq case.

shefty replied:
I think the sock_pe_entry is in need of a state machine overall.  :(  But to keep this solution simple for now... I don't think this requires an atomic.  The pe_entry must be protected against parallel access, and I think is in all cases.  This can be dropped back to u8 to match the other state variables (is_complete, mr_checked)

shefty commented on prov/sockets/src/sock_progress.c:
```diff
@@ -2534,18 +2534,19 @@ static void sock_pe_wait(struct sock_pe *pe)
 	CPU_ZERO(&mycpuset);
 
-	a = strtok_r(s, ",", &saveptra);
+	dup_s = strdup(s);
```
Need to check for strdup failure

sydidelot replied:
:+1:

shefty commented:
For the CM changes, are you using MSG EPs or RDM EPs?

shefty commented on prov/sockets/src/sock_epoll.c:
```diff
@@ -197,3 +197,114 @@ void sock_epoll_close(struct sock_epoll_set *set)
+		return -FI_ENOMEM;
+	}
+
+	if (socketpair(AF_UNIX, SOCK_STREAM, 0, emap->signal_fds) < 0)
```
The socket provider must be portable across Linux, FreeBSD, OS X, and Windows.

sydidelot replied:
Such a call to `socketpair()` is already present in Libfabric's master code, without distinction if the code is running on Linux, FreeBSD, OS X or Windows (see function `sock_conn_listen()` in `sock_conn.c`)

shefty commented on prov/sockets/include/sock.h:
```diff
@@ -1211,6 +1218,14 @@ int sock_epoll_wait(struct sock_epoll_set *set, int timeout);
+int sock_epoll_map_del(struct sock_epoll_map *emap, int fd);
+void *sock_epoll_map_lookup(struct sock_epoll_map *emap, int fd);
+int sock_epoll_map_signal(struct sock_epoll_map *emap);
+int sock_epoll_map_wait(struct sock_epoll_map *emap);
```
This appears generic and not tied specifically to the socket provider.  It would be better moved into general utility code, with a clear definition as to how it can be used.  The use case isn't clear from this patch.  Please see fi_util.h util_map abstraction to see if that provides functionality similar to what you're wanting, but at a higher level.

If you're just wanting to improve on the epoll portability abstraction, I would move to make that change to ofi_epoll itself, rather than creating a new abstraction over it.

sydidelot replied:
The basic idea behind this patch is to provide an interface that associates a pointer to a file descriptor returned by `epoll_wait`. This patch is a requirement for my big patch for the new connection manager and makes the code easier.

With epoll(7), you already can associate a pointer to a file descriptor. E.g:
`epoll_data_t evt;`
`evt.ptr = myHandle;`
`epoll_ctl(epoll_fd, EPOLL_CTL_ADD, fd, &evt) `
Then in `epoll_wait()` you can directly have access to `myHandle`.

While this feature is already implemented by `epoll`, Libfabric must support platforms where `epoll(7)` is not available but only `poll(2)` is. The problem is that `poll(2)` only returns a set of file descriptors, which is not convenient.

I've checked the file fi_util.h and I didn't find anything with similar functionalities.
If you prefer I can directly add my changes into the sock_epoll interface. Also, I think that the whole file `sockets/src/sock_epoll.c` should be moved into the general utility code.

shefty replied:
See include/fi_signal.h and src/common.c.  The fi_epoll abstraction should be providing the functionality that you need.  sock_epoll.c should be removed and replaced with the general implementation.

shefty replied:
I started a patch that converted sockets to use fi_epoll calls.  It was straightforward, except for the use in sock_pe_progress_rx_ep().  I couldn't follow what that call was doing exactly, or figure out if it was correct.  The way it's written, it looks like it could fail to progress a connection.  I don't have time to look into this in more detail at the moment.

Note that the fi_epoll_wait abstraction currently only returns information about the first event.  That functionality could be expanded to return multiple events if it is needed.

sydidelot replied:
You are right, the fi_epoll abstraction already provides the functionality I need for the epoll set.
I created a new PR (see PR #3221) that extracts the patches that are not directly related to the new connection manager for the sockets provider.
In this new PR, I have extended the fi_epoll_wait() call to allow multiple events to be returned.
I have also deprecated the the sock_epoll_* interface in favor of the common fi_epoll_* interface in the sockets provider.

sydidelot commented:
Sean,
Thank you very much for the review and for your feedback! 

Just to be sure I understand correctly the point about RxM:
RxM + TCP will emulate a support of Reliable Datagram over a reliable-connected protocol. One benefit of Reliable Datagram is that it doesn't require any connection establishment, so what's the reason of using MSG endpoints? Sorry for my basic question, I think I'm missing a piece of the puzzle :-)

I'm currently not familiar with RxM but I will probably have a look at the code in the next days. I've also figured out that there is a recent PR with the initial commit to the tcp provider. I'll keep a close eye to it!

sydidelot commented:
> For the CM changes, are you using MSG EPs or RDM EPs?
The code I've been using to test my CM changes is using MSG EPs only.

shefty commented:
There are providers that most naturally support connection-oriented protocols.  TCP is one.  Infiniband, RoCE, and iWarp are the others.  RXM will provide reliable-datagram semantics over those.  Note, that event though RXM uses connections underneath, it does not establish a connection until that peer is actually contacted.  It could be enhanced to disconnect if a connection remains idle for some amount of time.

There is also an RXD provider being developed.  That will provide reliable-datagram over DGRAM endpoints.  It would work with UDP and Infiniband UD communication.

If you're actually needing MSG EPs in your application, then I think the TCP provider will end up being a better fit (once it's actually written).  The sockets provider is architected around support for RDM EPs, with the other EPs being supported more or less as a by-product.

shefty commented:
Several of these patches have been merged.  Can you rebase the series over the tip of master and re-submit?  We can continue the discussion of the remaining patches.

sydidelot commented:
@shefty, I've just rebased my patch series on top of master. The pull request now includes only 5 patches.

shefty commented on prov/sockets/src/sock_progress.c:
```diff
@@ -192,12 +192,19 @@ static void sock_pe_release_entry(struct sock_pe *pe,
+{
+	memset(pe_entry, 0, sizeof(*pe_entry));
+	ofi_atomic_initialize32(&pe_entry->completion_reported, 0);
+}
```
This function then goes away

sydidelot replied:
Thanks! Change applied.

shefty commented on prov/sockets/src/sock_progress.c:
```diff
@@ -228,9 +235,19 @@ static struct sock_pe_entry *sock_pe_acquire_entry(struct sock_pe *pe)
 	int ret = 0;
+
+	if (sock_pe_is_completion_reported(pe_entry))
+		return;
```
This only protects against inserting multiple completions into the CQ. This check (and the one to report_recv_cq_completion) need to move up to the calling functions to ensure that the entire completion is only reported once.  Otherwise, we can bump the completion counter multiple times in cases where a counter is used in place of a CQ.

This means moving the check for ```if (pe_entry->completion_reported)``` into: sock_pe_report_send_completion(), sock_pe_report_recv_completion(), sock_pe_report_write_completion(), sock_pe_report_read_completion(), sock_pe_report_remote_read(), and wrapping around the call pair sock_pe_report_remote_write() + sock_pe_report_mr_completion().

I think that covers this, but we'll need to ensure that we get this right.

sydidelot replied:
Thanks, I've applied the modifications you have suggested. 
I created a new PR (see PR #3221) that extracts the patches that are not directly related to the new connection manager for the sockets provider.
We can continue the discussion here after PR #3221 get merged.

shefty commented on prov/sockets/src/sock_progress.c:
```diff
@@ -269,6 +286,10 @@ static void sock_pe_report_send_completion(struct sock_pe_entry *pe_entry)
 	int ret = 0;
+
+	if (sock_pe_is_completion_reported(pe_entry))
+		return;
```
see above comment

sydidelot replied:
Thanks! Change applied.

shefty commented on prov/sockets/include/sock.h:
```diff
@@ -167,6 +167,9 @@ enum {
 
 #define SOCK_WIRE_PROTO_VERSION (1)
 
+/* Forward declarations */
```
nit: comment is unnecessary

shefty commented on prov/sockets/include/sock.h:
```diff
@@ -536,6 +557,12 @@ struct sock_conn_handle {
 
+struct sock_ep_cm_entry {
+	int sock;
+	fastlock_t lock;
```
Can we avoid adding a lock per connection and use the cm_thread lock instead?

shefty commented on prov/sockets/include/sock.h:
```diff
@@ -907,12 +935,12 @@ struct sock_conn_hdr {
 	uint8_t type;
 	uint8_t reserved[3];
 	uint16_t port;
+	uint16_t data_len;
```
sock_conn_hdr is a wire protocol structure.  We can't modify it without breaking the protocol, which requires updating the version, handling backwards compatibility, etc.  This should be dropped.  If it is needed, it must be moved to a separate patch.

shefty commented on prov/sockets/include/sock.h:
```diff
@@ -907,12 +935,12 @@ struct sock_conn_hdr {
 };
 
 struct sock_conn_req {
-	struct sock_conn_hdr hdr;
```
I believe this is a wire protocol structure as well and should not be modified.

shefty commented on prov/sockets/include/sock.h:
```diff
@@ -923,20 +951,39 @@ enum {
 	SOCK_CONN_ACCEPT,
 	SOCK_CONN_REJECT,
 	SOCK_CONN_SHUTDOWN,
+	SOCK_CONN_FINALIZED,
```
These are wire protocol values.  Any changes break compatibility.

shefty commented on prov/sockets/include/sock.h:
```diff
@@ -923,20 +951,39 @@ enum {
+	uint16_t done_len;
+	uint8_t is_complete;
+	struct sock_ep_cm_handle *handle;
+};
```
Something isn't right with this.  It is mixing wire protocol structures (sock_conn_hdr) with non-wire fields (dlist_entry and handle).  Protocol structures should be kept separate from local structures unless there's a reason to copy a wire message to a local struct.

sydidelot commented:
@shefty Thanks for the review! I will work on new patches that preserve the wire protocol.

sydidelot commented:
@shefty, concerning your comment here:
> For new connections, in practice there will not be a need for the complexity of handling the case where a message cannot be sent because of insufficient buffering.

Shall we only consider blocking sockets for the connection manager, with the risk of blocking the current process in send()/recv() functions?

shefty commented:
No, I think we want non-blocking sockets.  I just don't know that we need to worry about being unable to send a single, small (< mtu) message immediately after the socket has been connected.

sydidelot commented:
Ok great, I'll modify the code accordingly.
I have another point for discussion: in the connection manager thread, when some data is ready on a file descriptor, shall we just call the recv() function and wait until all the whole message gets received? Or shall we (like in the patch) leave the recv() function if it returns EWOULDBLOCK (or if the message is partially received) and give it another attempt later?

shefty commented:
I'm okay blocking/spinning on sending/receiving the CM messages, once the socket indicates it's ready to send/receive.  In practice, anything less than an mtu should make it in a single go, and it doesn't seem worth the extra complexity to track CM messages between calls, like we need to for data transfers.

sydidelot commented:
Right, it makes sense. I'll simplify my code and submit a new patch. Thank you for your feedback!

sydidelot commented:
@shefty, I have taken into account your comments and I totally rewrote the code. As you requested, the patch doesn't break the wire protocol anymore and I removed the handling in case where a message cannot be sent because of insufficient buffering.
The patch still needs some adjustments but I first would like your general impression/feedback on these new patches.

Thanks a lot for your help!

shefty commented:
Please see the build failures for this PR.

sydidelot commented:
I found the issue in my patches and the PR finally passes the tests.
Thank you for your feedback!

shefty commented on prov/sockets/src/sock_conn.c:
```diff
@@ -384,25 +447,29 @@ int sock_conn_listen(struct sock_ep_attr *ep_attr)
+	fd_signal_set(&ep_attr->domain->conn_listener.signal);
+	fastlock_release(&ep_attr->domain->conn_listener.signal_lock);
+	if (ret)
+	{
```
nit: move { to same line as if (ret)

shefty commented on prov/sockets/include/sock.h:
```diff
@@ -204,6 +205,14 @@ struct sock_conn_map {
+struct sock_conn_listener {
+	fi_epoll_t emap;
+	struct fd_signal signal;
+	fastlock_t signal_lock;
```
Please add a comment that signal_lock must be acquired before map lock.

shefty commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -1236,3 +1256,158 @@ int sock_msg_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
+		ofi_close_socket(hreq->sock_fd);
+		hreq->sock_fd = -1;
+
+		pthread_mutex_lock(&hreq->mutex_cond.mutex);
```
fastlock converts to a spinlock.  We don't want to acquire a blocking mutex while holding a spinlock.

sydidelot replied:
Done

shefty commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -1236,3 +1256,158 @@ int sock_msg_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
 	return ret;
 }
 
+static void sock_ep_cm_check_msg_list(struct sock_ep_cm_head *cm_head)
```
Please rename call to indicate what is happening.  I.e. we're processing a close_list

sydidelot replied:
Done

shefty commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -1236,3 +1256,158 @@ int sock_msg_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
+
+				sock_ep_cm_unmonitor_handle(handle);
+				ofi_close_socket(handle->sock_fd);
+				handle->sock_fd = -1;
```
This duplicates code from sock_ep_cm_check_msg_list().  Use a common function or move the handle to the close_list and process there.

The pairing of sock_ep_cm_unmonitor_handle() and ofi_close_socket() occurs a few times, but not always.  It would be nice if this can be made consistent.

sydidelot replied:
Done. I've moved `ofi_close_socket()` into `sock_ep_cm_unmonitor_handle()`, which simplifies the code.

shefty commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -1078,6 +1092,12 @@ static int sock_pep_listen(struct fid_pep *pep)
 		return 0;
 
+	if (sock_ep_cm_start_thread(&_pep->cm_head))
+	{
```
nit: place { on same line as if ()

sydidelot replied:
Done

shefty commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -1236,3 +1256,158 @@ int sock_msg_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
+				continue;
+			}
+
+			if(sock_cm_recv(handle->sock_fd, &msg, sizeof(msg), MSG_PEEK)) {
```
We should just receive the message header and pass it to the handlers below.  This will save reading the same message twice.  The changes to sock_cm_recv() can be reverted then.

sydidelot replied:
Done. Thanks for the suggestion.

shefty commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -614,74 +634,62 @@ static int sock_ep_cm_connect(struct fid_ep *ep, const void *addr,
 	handle->req = req;
+	handle->cm_head = &_ep->attr->domain->cm_head;
+	pthread_mutex_init(&handle->mutex_cond.mutex, NULL);
+	pthread_cond_init(&handle->mutex_cond.cond, NULL);
```
It would be safe to initialize the handle when it's allocated.

sydidelot replied:
Done.

shefty commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -614,74 +634,62 @@ static int sock_ep_cm_connect(struct fid_ep *ep, const void *addr,
+	_ep->attr->info.handle = (void*) handle;
 	handle->ep = _ep;
 	handle->req = req;
+	handle->cm_head = &_ep->attr->domain->cm_head;
```
We can get to cm_head via handle->ep.  No need to duplicate the path.

sydidelot replied:
Done.

shefty commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -614,74 +634,62 @@ static int sock_ep_cm_connect(struct fid_ep *ep, const void *addr,
+		goto err;
+	}
+
+	if (sock_cm_send(sock_fd, req, sizeof(*req)))
```
The socket must be set to non-blocking.  Connect cannot guarantee to complete immediately.  We need async handling here prior to attempting to send on the socket.

sydidelot replied:
Thanks for the insightful comment. I have a separate patch that allows an asynchronous handling of the connection requests. While reading the socket provider, I've figured out that there is a similar issue with the function `sock_ep_connect` in `sock_conn.c` where `connect()` might block. We should probably also revisit this code (see https://github.com/ofiwg/libfabric/issues/2586).

shefty commented:
Both patches reviewed.  For the "prov/sockets: create one listener thread per domain" patch, the requested changes are straightforward.  I will pull in this patch manually, apply the necessary updates, and merge through a separate PR.

sydidelot commented on prov/sockets/include/sock.h:
```diff
@@ -490,24 +511,33 @@ struct sock_comp {
+enum {
+    SOCK_CM_STATE_DISCONNECTED = 0,
+    SOCK_CM_STATE_REQUESTED = 1,
+    SOCK_CM_STATE_CONNECTED = 2,
```
Done.
I just assigned the value 0 to SOCK_CM_STATE_DISCONNECTED and removed the other values.

sydidelot commented on prov/sockets/include/sock.h:
```diff
@@ -490,24 +511,33 @@ struct sock_comp {
-	uint64_t next_msg_id;
 	fastlock_t lock;
-	int is_connected;
+	int state;
```
Done.

sydidelot commented on prov/sockets/include/sock.h:
```diff
@@ -213,6 +213,16 @@ struct sock_conn_listener {
+	fi_epoll_t emap;
+	struct fd_signal signal;
+	fastlock_t signal_lock;
+	fastlock_t list_lock;
```
Done. I merged the `list_lock` into the `signal_lock`.

sydidelot commented on prov/sockets/include/sock.h:
```diff
@@ -213,6 +213,16 @@ struct sock_conn_listener {
+	pthread_t listener_thread;
+	struct dlist_entry close_list;
+	int do_listen;
+};
```
Your suggestion makes sense. I keep the idea for a future patch.

sydidelot commented on prov/sockets/include/sock.h:
```diff
@@ -500,15 +511,26 @@ struct sock_comp {
+
+struct sock_ep_cm_entry {
+	int sock;
+	fastlock_t lock;
```
Considering that this lock is per EP, I don't think we can reuse any existing lock. 
The best solution is probably to remove the lock and use an atomic CAS operation on the EP state. I've quickly checked `include/fi_atom.h` but I don't see any reference to CAS, though.

sydidelot commented on prov/sockets/include/sock.h:
```diff
@@ -500,15 +511,26 @@ struct sock_comp {
+struct sock_ep_cm_entry {
+	int sock;
+	fastlock_t lock;
+	int state;
```
Done

sydidelot commented on prov/sockets/include/sock.h:
```diff
@@ -910,12 +933,18 @@ enum {
 	int sock_fd;
+	int is_monitored;
 	int is_accepted;
+	int is_finalized;
```
Done. I have defined a structure with bit fields to store these 3 ints.

sydidelot commented on prov/sockets/include/sock.h:
```diff
@@ -910,12 +933,18 @@ enum {
+	struct {
+		pthread_mutex_t	mutex;
+		pthread_cond_t	cond;
+	} mutex_cond;
```
Done.

sydidelot commented on prov/sockets/src/sock_ep.c:
```diff
@@ -670,17 +670,24 @@ static int sock_ep_close(struct fid *fid)
+			pthread_mutex_unlock(&handle->mutex_cond.mutex);
+
+			free(handle->req);
+			free(handle);
```
Done.

sydidelot commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -1236,3 +1256,158 @@ int sock_msg_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
+				default:
+					SOCK_LOG_ERROR("Unexpected message type %d\n", msg.type);
+					break;
+			}
```
Done

sydidelot commented:
@shefty, thanks for your time reviewing my patches. I have considered all your comments and updated my patch series accordingly.
I have also included a patch to make fi_connect() asynchronous (commit c799074 " prov/sockets: asynchronously progress fi_connect()"). While this particular patch works fine with FreeBSD and Linux, it's constantly failing with the MacOSX's job in Travis. For this reason, I disabled this part of the code until I find the root cause of the issue.

shefty commented:
FYI - this PR hasn't been forgotten.  I've been on vacation, and with SC 17 coming up haven't had time to review the changes.  Also, the last update is causing crashes when running Sandia OpenSHMEM.  I'm trying to isolate the cause of that crash before integrating more changes to the socket provider.

lanl-ompi commented:
Can one of the admins verify this patch?

hppritcha commented:
add to whitelist

hppritcha commented:
the first run of LANL-CI will fail here because #3539 wasn't merged in to master when this CI test started.

hppritcha commented:
okay see if LANL CI is happy now.
bot:lanl:retest

lanl-ompi commented:
Test Passed

shefty commented on prov/sockets/include/sock.h:
```diff
@@ -905,15 +924,29 @@ enum {
+			uint8_t rejected:  1;
+			uint8_t closing:   1;
+		};
+	} flags;
```
Looking at how these are used, we can make the following changes:
- monitored: This looks okay as a flag.
- accepted: This can be removed.  It's only used as a minor validation check against the user accepting then rejecting the same request.

The last 3 are independent and are better as states than flags.  A connection is either rejected or closing, but not both.  And finalized is reached by going through closing.  Accepted/active could be kept as another state used prior to moving to closing or rejected.

sydidelot replied:
Thanks for the review!
I removed the list of flags. Instead I used an enum to list all the possible states a connection handler can have.

shefty commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -1235,3 +1341,154 @@ int sock_msg_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
+		if (handle->ep) {
+			if (!handle->pep) /* Active-side */
+				sock_ep_cm_report_connect_fail(handle->ep, NULL, 0);
+			sock_ep_cm_shutdown_report(handle->ep, 0);
```
Reporting a connection failure and shutdown event for the same connection request looks wrong.

sydidelot replied:
Done. I rewrote this part accordingly.

shefty commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -894,23 +947,117 @@ static struct fi_info *sock_ep_msg_get_info(struct sock_pep *pep,
+void sock_ep_cm_signal_locked(struct sock_ep_cm_head *cm_head)
+{
+	fd_signal_set(&cm_head->signal);
+}
```
Drop this function and just call fd_signal_set() directly.

sydidelot commented on prov/sockets/src/sock_ep.c:
```diff
@@ -670,17 +670,14 @@ static int sock_ep_close(struct fid *fid)
-				       ofi_syserr());
+		handle = container_of(sock_ep->attr->info.handle,
+		                      struct sock_conn_req_handle, handle);
+		if (handle) {
```
Oh right, I fixed the code.

sydidelot commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -432,106 +432,149 @@ static int sock_cm_recv(int fd, void *buf, int len)
+                                      uint32_t events)
+{
+	fastlock_acquire(&cm_head->signal_lock);
+	sock_ep_cm_monitor_handle_locked(cm_head, handle, callback, events);
```
I simplified the code and removed the useless callback function.

sydidelot commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -432,106 +432,149 @@ static int sock_cm_recv(int fd, void *buf, int len)
+	if (close_socket && handle->sock_fd >= 0)
+	{
+		ofi_close_socket(handle->sock_fd);
+		handle->sock_fd = -1;
```
Done. I also took the opportunity of this patch to fix the same issue in some other places.

sydidelot commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -1235,3 +1341,154 @@ int sock_msg_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
+
+			/* The handle has been removed from epoll since the last call to
+			 * fi_epoll_wait() */
+			if (!handle->flags.monitored)
```
If `handle->flags.monitored != 1`, it's likely an invalid memory access. I removed the condition and used a call to assert() instead.

sydidelot commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -1235,3 +1341,154 @@ int sock_msg_passive_ep(struct fid_fabric *fabric, struct fi_info *info,
+				continue;
+
+			assert(handle->event_callback);
+			handle->event_callback(cm_head, handle);
```
Done, I simplified the code and removed the useless callback function.

sydidelot commented:
I have modified the code according to your comments.

Please note that I removed the last patch that attempted to make the call to `connect()` asynchronous (this patch was not working anyway). I will address this issue later if you don't mind.

hppritcha commented:
I'm doing an experiment with the lanl jenkins using this pr.  please ignore if it fails.
bot:lanl:retest

sydidelot commented:
@shefty, I apologize for my very long silence concerning the pull request, but I had a very busy beginning of year 2018. I just rebased my patch and applied the code modifications you suggested later. 

@vkrishna, I see you've contributed a lot to the TCP provider. Could you please review my changes to the socket provider when you have time? There is probably some parts of the code that overlap with your contribution to the TCP provider. Thanks!

sydidelot commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -431,106 +431,146 @@ static int sock_cm_recv(int fd, void *buf, int len)
-		}
+	handle->monitored = 1;
+	int ret = fi_epoll_add(cm_head->emap, handle->sock_fd,
+	                       events, handle);
```
Done

sydidelot commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -431,106 +431,146 @@ static int sock_cm_recv(int fd, void *buf, int len)
 
-		if (msg.type == SOCK_CONN_SHUTDOWN)
-			break;
+	int ret = fi_epoll_del(cm_head->emap, handle->sock_fd);
```
Done

sydidelot commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -431,106 +431,146 @@ static int sock_cm_recv(int fd, void *buf, int len)
+	 * same time. Some caution is required to prevent a socket from being
+	 * close concurrently, which could cause an unexpected socket to be
+	 * closed by mistake. */
+	if (close_socket && handle->sock_fd >= INVALID_SOCKET)
```
Good catch! Thanks!

sydidelot commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -431,106 +431,146 @@ static int sock_cm_recv(int fd, void *buf, int len)
+	}
+	else if (ep->attr->cm.state == SOCK_CM_STATE_REQUESTED) {
+		ep->attr->cm.state = SOCK_CM_STATE_DISCONNECTED;
 	}
```
Done

sydidelot commented on prov/sockets/src/sock_ep_msg.c:
```diff
@@ -431,106 +431,146 @@ static int sock_cm_recv(int fd, void *buf, int len)
+					 FI_ECONNREFUSED, -FI_ECONNREFUSED,
+					 NULL, 0))
+			SOCK_LOG_ERROR("Error in writing to EQ\n");
 	}
```
Done

vkrishna commented:
@sydidelot Sorry for not doing it earlier. I will check your PR. Thanks

vkrishna commented:
It looks good.  

There is quite a bit of overlap in connection management part between tcp and sockets.  There is good potential to break it into manageable parts and push it to util code for maintainability. It is not practical to do this within this PR, considering tcp provider is not complete as of now.

sydidelot commented:
@shefty Thanks for merging my patch and taking the time to review it.
Please let me know if the new code is not working as excepted: I can help with debugging.

shefty commented:
Thanks - at the moment all tests in travis and our internal CI are passing with these changes.  So things look good so far!
